### PR TITLE
refactor(proto): unify protobuf types to represent column index and column ref

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Remove unused `type` field of `AggCall::OrderByField` (will remove this type later)
- Use `optional uint32` to replace `catalog.ColumnIndex` for `row_id_index` and `vnode_col_index` (optional column)
- Use `uint32` for required column index
- Add `ColumnRef` type to replace `AggCall::Arg` and `Watermark::col_idx` + `Watermark::data_type`

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: proto/expr.proto [hunk 1]
```diff
@@ -141,7 +141,7 @@ message ExprNode {
   Type expr_type = 1;
   data.DataType return_type = 3;
   oneof rex_node {
-    InputRefExpr input_ref = 4;
+    uint32 input_ref = 4;
```
Comment (BugenZhao) [inline proto/expr.proto:144]: Not sure whether this is a good change. 🤔 With a new type of `InputRefExpr`, the meaning of this unsigned integer seems more explicit.

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM!

File: proto/expr.proto (hunk 1 above)
Comment (richardchien) [reply proto/expr.proto:144]: The rationale here is to uniformly use `uint32` as column index in proto. Previously we have `InputRefExpr`, but it's different from `InputRef` and `InputRefExpression` in Rust code. And we have another `Arg` type containing what `InputRef` has, which is confusing.

I tried to add a `ColumnIndex` type in proto, but found it hard to choose between `ColumnIndex` and `uint32` in many other proto message defs. E.g. we have `uint32 row_count_index` field referencing one column, we also have `repeated uint32 output_indices` referencing a vector of columns, if we use `ColumnIndex` for the former, should we also use it for the latter? If we do use it for all column index fields, then too many code need to be modified (and they will be `None`-able); if we don't, then we will face some inconsistency and `ColumnIndex` won't be able to hide the unsigned integer type.

File: proto/expr.proto [hunk 2]
```diff
@@ -161,8 +161,10 @@ message TableFunction {
   data.DataType return_type = 3;
 }
 
-message InputRefExpr {
-  int32 column_idx = 1;
+// Reference to a column, containing its index and data type.
+message ColumnRef {
```
Comment (st1page) [inline proto/expr.proto:165]: prefer `InputRef`. sometimes someone mixed up the input_col_idx and output_col_idx.

**Review by st1page — APPROVED** [review]
LGTM.
BTW, there is also a mess for our "key" in proto which usually is int32

**Pull request merged**